src/vs/workbench/contrib/tasks/common/taskService.ts

```diff
@@ -13,14 +13,16 @@ import { IWorkspaceFolder, IWorkspace } from 'vs/platform/workspace/common/works
 import { Task, ContributedTask, CustomTask, ITaskSet, TaskSorter, ITaskEvent, ITaskIdentifier, ConfiguringTask, TaskRunSource } from 'vs/workbench/contrib/tasks/common/tasks';
 import { ITaskSummary, ITaskTerminateResponse, ITaskSystemInfo } from 'vs/workbench/contrib/tasks/common/taskSystem';
 import { IStringDictionary } from 'vs/base/common/collections';
-import { RawContextKey } from 'vs/platform/contextkey/common/contextkey';
+import { RawContextKey, ContextKeyExpr } from 'vs/platform/contextkey/common/contextkey';
 
 export { ITaskSummary, Task, ITaskTerminateResponse as TaskTerminateResponse };
 
-export const CustomExecutionSupportedContext = new RawContextKey<boolean>('customExecutionSupported', true, nls.localize('tasks.customExecutionSupported', "Whether CustomExecution tasks are supported. Consider using in the when clause of a \'taskDefinition\' contribution."));
+export const CustomExecutionSupportedContext = new RawContextKey<boolean>('customExecutionSupported', false, nls.localize('tasks.customExecutionSupported', "Whether CustomExecution tasks are supported. Consider using in the when clause of a \'taskDefinition\' contribution."));
 export const ShellExecutionSupportedContext = new RawContextKey<boolean>('shellExecutionSupported', false, nls.localize('tasks.shellExecutionSupported', "Whether ShellExecution tasks are supported. Consider using in the when clause of a \'taskDefinition\' contribution."));
 export const TaskCommandsRegistered = new RawContextKey<boolean>('taskCommandsRegistered', false, nls.localize('tasks.taskCommandsRegistered', "Whether the task commands have been registered yet"));
 export const ProcessExecutionSupportedContext = new RawContextKey<boolean>('processExecutionSupported', false, nls.localize('tasks.processExecutionSupported', "Whether ProcessExecution tasks are supported. Consider using in the when clause of a \'taskDefinition\' contribution."));
+export const ServerlessWebContext = new RawContextKey<boolean>('serverlessWebContext', false, nls.localize('tasks.serverlessWebContext', "True when in the web with no remote authority."));
+export const TaskExecutionSupportedContext = ContextKeyExpr.or((ContextKeyExpr.and(ShellExecutionSupportedContext, ProcessExecutionSupportedContext, ServerlessWebContext.toNegated()), (ContextKeyExpr.and(ServerlessWebContext, CustomExecutionSupportedContext))));
```

Comment:
Shouldn't this be this since one of those are required to have tasks work? 

```ts
ContextKeyExpr.and(ShellExecutionSupportedContext, ProcessExecutionSupportedContext, CustomExecutionSupportedContext)
```

Now sure what `CustomExecutionSupportedContext` does atm but it should probably be `!isWeb || remoteAuthority`?

Reply:
We require custom execution (extension provided) type for serverless and process or shell type for the other case

Reply:
ServerlessWebContext is the part that does remote authority and web

Reply:
this shouldn't be set to true on web unless there are extension contributed tasks - which is when `CustomExecutionSupportedContext` gets set to true

Reply:
so

`serverlessWeb & extensionContributedTask or not serverless web & process/shell tasks`

Reply:
`ServerlessWebContext` = `isWeb && !remoteAuthority`. I just looked and was wrong abt what `CustomExecutionSupportedContext` does. I think we need to check that there's a provider for the serverless web context

Reply:
pretty sure that only `customExecution` tasks are supported on serverless web

Reply:
Yeah, instead of bringing the environment into it here though, wouldn't it be simplest to just check if the different types are supported?

Reply:
oh I see what you mean. Not sure why `customExecution` wasn't included in the original one and don't want to break things bc think that was probably by design

meaning before the context key only checked process or shell types

Reply:
Oh I mean or in my example above:

```ts
ContextKeyExpr.or(ShellExecutionSupportedContext, ProcessExecutionSupportedContext, CustomExecutionSupportedContext)
```

Provided `CustomExecutionSupportedContext` is only true when there's a provider it seems like the most reliable way to do it.

Since before it was using this:

```ts
const SHOW_TASKS_COMMANDS_CONTEXT = ContextKeyExpr.and(ShellExecutionSupportedContext, ProcessExecutionSupportedContext);
```

How about this instead?

```ts
ContextKeyExpr.or(ContextKeyExpr.and(ShellExecutionSupportedContext, ProcessExecutionSupportedContext), CustomExecutionSupportedContext);
```

Reply:
sure - not sure that simplifies it much since it's removing part of the conditional that's pretty important IMO. 

so will now instead of checking `TaskExecutionSupportedContext` will need to check `!ServerlessWebContext || TaskExecutionSupportedContext` which is a bit confusing

Reply:
went with that as hoping to get this merged ASAP as it's pretty bad atm